File: content/develop/use-cases/semantic-cache/nodejs/embeddings.js
// Local text-embedding helper backed by @xenova/transformers.
//
// This is a thin wrapper around the ONNX-exported sentence-transformers
// model `Xenova/all-MiniLM-L6-v2`: a 384-dimensional encoder that runs
// in-process on CPU through ONNX Runtime Web, needs no API key, and
// produces vectors that are numerically very close to the equivalent
// PyTorch model (close enough that paraphrase distances differ only at
// the fourth decimal place — see the smoke-test in the README).
//
// Vectors are L2-normalised so a Redis Search index declared with
// `DISTANCE_METRIC COSINE` returns scores that are directly comparable
// across entries. The model is downloaded into the local Hugging Face
// cache on the first call; every later call runs offline.

import { env, pipeline } from '@xenova/transformers';

// Allow the local cache to satisfy subsequent runs without re-downloading.
env.allowLocalModels = true;

const DEFAULT_MODEL = 'Xenova/all-MiniLM-L6-v2';

export class LocalEmbedder {
  // Use `LocalEmbedder.create(...)` instead of `new LocalEmbedder(...)`
  // because the pipeline load is async; we want one place that owns
  // the wait and the dimension probe.
  constructor(modelName, extractor, dim) {
    this.modelName = modelName;
    this.extractor = extractor;
    this.dim = dim;
  }

  static async create(modelName = DEFAULT_MODEL) {
    const extractor = await pipeline('feature-extraction', modelName);
    // Probe the output shape once and record it on the instance so
    // callers can compare against the cache's expected vectorDim
    // before doing any inserts. RedisSemanticCache also checks
    // length on every put / lookup, so a model swap that produces
    // wrong-dim vectors fails at the call site with a clear error.
    const probe = await extractor('dimension probe', {
      pooling: 'mean', normalize: true,
    });
    const dim = probe.dims[probe.dims.length - 1];
    return new LocalEmbedder(modelName, extractor, dim);
  }

  // Encode a single string. Returns a Float32Array of length `dim`.
  async encodeOne(text) {
    const out = await this.extractor(text, {
      pooling: 'mean', normalize: true,
    });
    return new Float32Array(out.data);
  }

  // Encode several strings in one pipeline call. Returns an array of
  // Float32Array; callers that need raw bytes use `toBytes` per row.
  async encodeMany(texts) {
    const out = await this.extractor(texts, {
      pooling: 'mean', normalize: true,
    });
    const rows = out.dims[0];
    const cols = out.dims[1];
    const result = [];
    for (let i = 0; i < rows; i++) {
      result.push(new Float32Array(out.data.slice(i * cols, (i + 1) * cols)));
    }
    return result;
  }

  // Pack a Float32Array into the bytes Redis Search expects.
  // Float32Array.buffer is little-endian on every architecture we care
  // about — Node runs on x86_64/arm64, both little-endian.
  static toBytes(vector) {
    return Buffer.from(vector.buffer, vector.byteOffset, vector.byteLength);
  }
}

File: content/develop/use-cases/semantic-cache/redis-py/embeddings.py
"""
Local text-embedding helper backed by sentence-transformers.

This is a thin wrapper around the ``sentence-transformers`` model
``all-MiniLM-L6-v2``: a 384-dimensional encoder that runs on CPU,
needs no API key, and has a small footprint (~80 MB). On the first
call the model is downloaded into the local Hugging Face cache; every
later call runs locally.

Vectors are L2-normalised on output so a Redis Search index declared
with ``DISTANCE_METRIC COSINE`` returns scores that are directly
comparable across entries.
"""

from __future__ import annotations

from typing import Iterable

import numpy as np


_DEFAULT_MODEL = "sentence-transformers/all-MiniLM-L6-v2"


class LocalEmbedder:
    """Encode short strings into normalised float32 vectors.

    A single instance loads the model once and reuses it for every
    call. The demo server keeps one ``LocalEmbedder`` around for the
    lifetime of the process so it can embed the incoming prompt and,
    on a miss, write the same bytes into the cache without
    re-encoding.
    """

    def __init__(self, model_name: str = _DEFAULT_MODEL) -> None:
        from sentence_transformers import SentenceTransformer

        self.model_name = model_name
        self.model = SentenceTransformer(model_name)
        self.dim = int(self.model.get_sentence_embedding_dimension())

    def encode_one(self, text: str) -> np.ndarray:
        """Encode a single string. Returns a 1-D ``float32`` array."""
        return self.encode_many([text])[0]

    def encode_many(self, texts: Iterable[str]) -> np.ndarray:
        """Encode a batch. Returns an ``(N, dim) float32`` array."""
        batch = list(texts)
        vectors = self.model.encode(
            batch,
            batch_size=32,
            normalize_embeddings=True,
            convert_to_numpy=True,
        )
        return vectors.astype(np.float32, copy=False)
